File: app/src/main/java/org/thoughtcrime/securesms/recipients/ui/sharablegrouplink/ShareableGroupLinkViewModel.java
package org.thoughtcrime.securesms.recipients.ui.sharablegrouplink;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import org.thoughtcrime.securesms.groups.GroupId;
import org.thoughtcrime.securesms.groups.LiveGroup;
import org.thoughtcrime.securesms.groups.ui.GroupChangeFailureReason;
import org.thoughtcrime.securesms.groups.ui.GroupErrors;
import org.thoughtcrime.securesms.groups.v2.GroupLinkUrlAndStatus;
import org.thoughtcrime.securesms.util.AsynchronousCallback;
import org.thoughtcrime.securesms.util.SingleLiveEvent;

final class ShareableGroupLinkViewModel extends ViewModel {

  private final ShareableGroupLinkRepository    repository;
  private final LiveData<GroupLinkUrlAndStatus> groupLink;
  private final SingleLiveEvent<Integer>        toasts;
  private final SingleLiveEvent<Boolean>        busy;
  private final LiveData<Boolean>               canEdit;

  private ShareableGroupLinkViewModel(@NonNull GroupId.V2 groupId, @NonNull ShareableGroupLinkRepository repository) {
    LiveGroup liveGroup = new LiveGroup(groupId);

    this.repository = repository;
    this.groupLink  = liveGroup.getGroupLink();
    this.canEdit    = liveGroup.isSelfAdmin();
    this.toasts     = new SingleLiveEvent<>();
    this.busy       = new SingleLiveEvent<>();
  }

  LiveData<GroupLinkUrlAndStatus> getGroupLink() {
    return groupLink;
  }

  LiveData<Integer> getToasts() {
    return toasts;
  }

  LiveData<Boolean> getBusy() {
    return busy;
  }

  LiveData<Boolean> getCanEdit() {
    return canEdit;
  }

  void onToggleGroupLink() {
    busy.setValue(true);
    repository.toggleGroupLinkEnabled(new AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason>() {
      @Override
      public void onComplete(@Nullable Void result) {
        busy.postValue(false);
      }

      @Override
      public void onError(@Nullable GroupChangeFailureReason error) {
        busy.postValue(false);
        toasts.postValue(GroupErrors.getUserDisplayMessage(error));
      }
    });
  }

  void onToggleApproveMembers() {
    busy.setValue(true);
    repository.toggleGroupLinkApprovalRequired(new AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason>() {
      @Override
      public void onComplete(@Nullable Void result) {
        busy.postValue(false);
      }

      @Override
      public void onError(@Nullable GroupChangeFailureReason error) {
        busy.postValue(false);
        toasts.postValue(GroupErrors.getUserDisplayMessage(error));
      }
    });
  }

  void onResetLink() {
    busy.setValue(true);
    repository.cycleGroupLinkPassword(new AsynchronousCallback.WorkerThread<Void, GroupChangeFailureReason>() {
      @Override
      public void onComplete(@Nullable Void result) {
         busy.postValue(false);
      }

      @Override
      public void onError(@Nullable GroupChangeFailureReason error) {
        busy.postValue(false);
        toasts.postValue(GroupErrors.getUserDisplayMessage(error));
      }
    });
  }

  public static final class Factory implements ViewModelProvider.Factory {

    private final GroupId.V2                   groupId;
    private final ShareableGroupLinkRepository repository;

    public Factory(@NonNull GroupId.V2 groupId, @NonNull ShareableGroupLinkRepository repository) {
      this.groupId    = groupId;
      this.repository = repository;
    }

    @Override
    public @NonNull <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
      //noinspection ConstantConditions
      return modelClass.cast(new ShareableGroupLinkViewModel(groupId, repository));
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/recipients/ui/sharablegrouplink/ShareableGroupLinkViewModel.kt
/*
 * Copyright 2026 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.thoughtcrime.securesms.recipients.ui.sharablegrouplink

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import org.signal.chatsettings.screens.sharablegrouplink.ShareableGroupLinkEvents
import org.signal.chatsettings.screens.sharablegrouplink.ShareableGroupLinkState
import org.signal.chatsettings.screens.sharablegrouplink.ShareableGroupLinkState.Dialog
import org.signal.core.ui.compose.EventDrivenViewModel
import org.signal.core.util.logging.Log
import org.thoughtcrime.securesms.groups.GroupId
import org.thoughtcrime.securesms.groups.ui.GroupChangeResult
import org.thoughtcrime.securesms.groups.ui.GroupErrors

/**
 * View model behind [ShareableGroupLinkScreen].
 */
class ShareableGroupLinkViewModel(
  private val groupId: GroupId.V2,
  private val repository: ShareableGroupLinkRepository = ShareableGroupLinkRepository()
) : EventDrivenViewModel<ShareableGroupLinkEvents>(TAG, false) {

  companion object {
    private val TAG = Log.tag(ShareableGroupLinkViewModel::class)
  }

  private val _state = MutableStateFlow(ShareableGroupLinkState())

  val state: StateFlow<ShareableGroupLinkState> = _state.asStateFlow()

  init {
    repository
      .observeGroupLink(groupId)
      .onEach { onEvent(ShareableGroupLinkEvents.GroupLinkChanged(it)) }
      .launchIn(viewModelScope)
  }

  override suspend fun processEvent(event: ShareableGroupLinkEvents) {
    when (event) {
      is ShareableGroupLinkEvents.GroupLinkChanged -> {
        _state.update { it.copy(groupLink = event.groupLink) }
      }

      ShareableGroupLinkEvents.GroupLinkToggled -> {
        val current = _state.value.groupLink

        applyChange {
          repository.setGroupLinkState(
            groupId = groupId,
            enabled = !current.enabled,
            requiresAdminApproval = current.requiresAdminApproval
          )
        }
      }

      ShareableGroupLinkEvents.AdminApprovalToggled -> {
        val current = _state.value.groupLink

        applyChange {
          repository.setGroupLinkState(
            groupId = groupId,
            enabled = current.enabled,
            requiresAdminApproval = !current.requiresAdminApproval
          )
        }
      }

      ShareableGroupLinkEvents.ResetLinkClicked -> {
        _state.update { it.copy(dialog = Dialog.ConfirmResetLink) }
      }

      ShareableGroupLinkEvents.ResetLinkConfirmed -> {
        _state.update { it.copy(dialog = Dialog.None) }
        applyChange { repository.cycleGroupLinkPassword(groupId) }
      }

      ShareableGroupLinkEvents.DialogDismissed -> {
        _state.update { it.copy(dialog = Dialog.None) }
      }

      ShareableGroupLinkEvents.SnackbarDismissed -> {
        _state.update { it.copy(errorMessage = null) }
      }
    }
  }

  /**
   * Runs a group change while the screen shows a spinner over itself. Changes have to be applied one at a time, since
   * each toggle is relative to the link's current state.
   */
  private suspend fun applyChange(change: suspend () -> GroupChangeResult) {
    _state.update { it.copy(busy = true) }

    val result = change()

    _state.update { it.copy(busy = false) }

    if (!result.isSuccess) {
      _state.update { it.copy(errorMessage = GroupErrors.getUserDisplayMessage(result.failureReason)) }
    }
  }
}
